# Assignment Sesi 9
Nama: Faraday Barr Fatahillah

Instruksi

Anda adalah data analyst junior di sebuah perusahaan riset maritim. Tim Anda sedang menyusun laporan historis tentang insiden Titanic dan membutuhkan beberapa angka agregasi dari data penumpang.

Ini data dari seaborn atau dari kaggle? karena beda
1. seabron : https://github.com/mwaskom/seaborn-data/blob/master/titanic.csv
2. Kaggle  : https://www.kaggle.com/datasets/zhangxuanxuan/titanicdatasets


**Untuk tugas ini menyesuaikan dengan instruksi dari Seaborn langsung**

**Task 1 — Eksplorasi awal DataFrame**

1. Load dataset Titanic ke variabel df.
2. Cetak: jumlah baris dan kolom, nama kolom beserta tipe datanya, dan jumlah nilai null per kolom (urutkan dari terbanyak).
3. Tampilkan statistik deskriptif hanya untuk kolom numerik menggunakan describe(). Identifikasi secara tertulis (di cell markdown): kolom mana yang punya nilai minimum atau maksimum yang terlihat tidak wajar?

In [49]:
import seaborn as sns
import pandas as pd

#Task 1.1
df = sns.load_dataset('titanic')

#Task 1.2
print(f'Jumlah Baris: {len(df)}')
print(f'Jumlah kolom: {len(df.columns)}')
print(df.shape)

print('')
print('Nama kolom dan tipe data:')
df.info()

print('')
print('Jumlah nilai null per kolom:')
df.isnull().sum().sort_values(ascending=False)

Jumlah Baris: 891
Jumlah kolom: 15
(891, 15)

Nama kolom dan tipe data:
<class 'pandas.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 15 columns):
 #   Column       Non-Null Count  Dtype   
---  ------       --------------  -----   
 0   survived     891 non-null    int64   
 1   pclass       891 non-null    int64   
 2   sex          891 non-null    str     
 3   age          714 non-null    float64 
 4   sibsp        891 non-null    int64   
 5   parch        891 non-null    int64   
 6   fare         891 non-null    float64 
 7   embarked     889 non-null    str     
 8   class        891 non-null    category
 9   who          891 non-null    str     
 10  adult_male   891 non-null    bool    
 11  deck         203 non-null    category
 12  embark_town  889 non-null    str     
 13  alive        891 non-null    str     
 14  alone        891 non-null    bool    
dtypes: bool(2), category(2), float64(2), int64(4), str(5)
memory usage: 80.7 KB

Jumlah nilai null per

deck           688
age            177
embarked         2
embark_town      2
sex              0
pclass           0
survived         0
fare             0
parch            0
sibsp            0
class            0
adult_male       0
who              0
alive            0
alone            0
dtype: int64

In [50]:
#Task 1.3
df.describe()

,survived,pclass,age,sibsp,parch,fare
count,891.000000,891.000000,714.000000,891.000000,891.000000,891.000000
mean,0.383838,2.308642,29.699118,0.523008,0.381594,32.204208
std,0.486592,0.836071,14.526497,1.102743,0.806057,49.693429
min,0.000000,1.000000,0.420000,0.000000,0.000000,0.000000
25%,0.000000,2.000000,20.125000,0.000000,0.000000,7.910400
50%,0.000000,3.000000,28.000000,0.000000,0.000000,14.454200
75%,1.000000,3.000000,38.000000,1.000000,0.000000,31.000000
max,1.000000,3.000000,80.000000,8.000000,6.000000,512.329200


Answer:
Nilai minimum dan maksimum paling adalah 
1. age karena memiliki nilai minimum 0.42 selain itu datanya juga tidak full (hanya ada 714 data dari 891)
2. fare karena nilai minimum 0 yang seharusnya ada harga dan juga nilai maksimum sebesar 512.3292 karena harga tiket itu mahal sekali untuk tahun 1900-an

**Task 2 — Querying dengan kondisi majemuk**
1. Filter penumpang yang memenuhi SEMUA kriteria berikut sekaligus: kelas kabin 1 atau 2, usia di atas 30 tahun, dan tidak selamat (survived = 0). Cetak jumlah penumpang  yang memenuhi kriteria tersebut.
2. Filter penumpang perempuan yang selamat dan berusia di bawah 18 tahun. Tampilkan kolom: name, age, pclass, embark_town.
3. Buat kolom baru bernama kelompok_usia dengan kategori berikut:    - 'Anak'    : usia < 12    - 'Remaja' : usia 12 sampai 17    - 'Dewasa' : usia 18 sampai 59    - 'Lansia' : usia >= 60    Gunakan pd.cut() atau np.where() bertingkat. Baris dengan usia null dibiarkan NaN.

In [51]:
#Task 2.1
df_filter1 = df.query('pclass < 3 and age > 30 and survived == 0')
print(f'Filter 1: {len(df_filter1)}')

Filter 1: 94


In [52]:
#Task 2.2
print('')
print('Filter 2:') 
print(df.query('sex == "female" and age <= 18.0')[['age', 'pclass', 'embark_town']])



Filter 2:
      age  pclass  embark_town
9    14.0       2    Cherbourg
10    4.0       3  Southampton
14   14.0       3  Southampton
22   15.0       3   Queenstown
24    8.0       3  Southampton
..    ...     ...          ...
830  15.0       3    Cherbourg
852   9.0       3    Cherbourg
853  16.0       1  Southampton
855  18.0       3  Southampton
875  15.0       3    Cherbourg

[68 rows x 3 columns]


In [53]:
#Task 2.3
print('')
print('Dataframe dengan kolom baru')
bins = [0, 12, 17, 60, float('inf')]
label = ['Anak', 'Remaja', 'Dewasa', 'Lansia']
df['kelompok_usia'] = pd.cut(df['age'], bins=bins, labels=label)
df.head(10)


Dataframe dengan kolom baru


,survived,pclass,sex,age,sibsp,parch,fare,embarked,class,who,adult_male,deck,embark_town,alive,alone,kelompok_usia
0,0,3,male,22.0,1,0,7.2500,S,Third,man,True,NaN,Southampton,no,False,Dewasa
1,1,1,female,38.0,1,0,71.2833,C,First,woman,False,C,Cherbourg,yes,False,Dewasa
2,1,3,female,26.0,0,0,7.9250,S,Third,woman,False,NaN,Southampton,yes,True,Dewasa
3,1,1,female,35.0,1,0,53.1000,S,First,woman,False,C,Southampton,yes,False,Dewasa
4,0,3,male,35.0,0,0,8.0500,S,Third,man,True,NaN,Southampton,no,True,Dewasa
5,0,3,male,NaN,0,0,8.4583,Q,Third,man,True,NaN,Queenstown,no,True,NaN
6,0,1,male,54.0,0,0,51.8625,S,First,man,True,E,Southampton,no,True,Dewasa
7,0,3,male,2.0,3,1,21.0750,S,Third,child,False,NaN,Southampton,no,False,Anak
8,1,3,female,27.0,0,2,11.1333,S,Third,woman,False,NaN,Southampton,yes,False,Dewasa
9,1,2,female,14.0,1,0,30.0708,C,Second,child,False,NaN,Cherbourg,yes,False,Remaja


**Task 3 — Groupby dan agregasi**

1. Hitung survival rate (persen penumpang yang selamat) per kelas kabin (pclass). Tampilkan hasilnya sebagai DataFrame dengan kolom pclass, jumlah_penumpang, jumlah_selamat, dan survival_rate (dalam persen, 2 desimal).
2. Gunakan groupby dengan agg() untuk menghitung rata-rata usia dan rata-rata tarif (fare) sekaligus, dikelompokkan berdasarkan pclass dan sex. Tampilkan hasilnya.
3. Dari kolom kelompok_usia yang dibuat di Task 2, hitung jumlah penumpang yang selamat dan tidak selamat per kelompok usia. Tampilkan sebagai tabel pivot menggunakan pd.crosstab() atau groupby.

In [54]:
#Task 3.1
survive_rate = df.groupby('pclass').agg(
    jumlah_selamat=('survived', 'sum'),
    jumlah_penumpang=('survived', 'count') 
).reset_index()

survive_rate['survival_rate'] = (survive_rate['jumlah_selamat'] / survive_rate['jumlah_penumpang']*100).round(2)

survive_rate

,pclass,jumlah_selamat,jumlah_penumpang,survival_rate
0,1,136,216,62.96
1,2,87,184,47.28
2,3,119,491,24.24


In [55]:
#Task 3.2
fare_age = df.groupby(['pclass', 'sex']).agg(
    avg_age=('age', 'mean'),
    avg_fare=('fare', 'mean') 
).round(2).reset_index()

fare_age

,pclass,sex,avg_age,avg_fare
0,1,female,34.61,106.13
1,1,male,41.28,67.23
2,2,female,28.72,21.97
3,2,male,30.74,19.74
4,3,female,21.75,16.12
5,3,male,26.51,12.66


In [ ]:
#Task 3.3
pivot_table = pd.crosstab(df['```kelompok_usia'], df['survived'], colnames=['survived'])
pivot_table.columns = ['tidak_selamat', 'selamat']

pivot_table

,tidak_selamat,selamat
kelompok_usia,,
Anak,29,40
Remaja,23,21
Dewasa,355,224
Lansia,17,5


**Task 4 — Interpretasi**

Di cell markdown terpisah, tulis minimal 3 kalimat interpretasi berdasarkan hasil Task 3. Setiap kalimat harus menyebut angka spesifik dari hasil agregasi, bukan hanya pernyataan umum.
Contoh format yang diharapkan: "Penumpang kelas 1 memiliki survival rate tertinggi sebesar X%, jauh di atas kelas 3 yang hanya Y%."

Answer:
Hasil yang saya dapat dari data agregasi di atas adalah semakin banyak uang yang dimiliki, semakin besar kesempatan untuk selamat. Hal ini bisa di lihat yang pertama adalah class 1 memiliki rata-rata harga tiket yang sangat tinggi seperti pada perempuan sebesar 106.13 dan laki-laki sebesar 67.23. Perbedaan ini sangat jauh dengan kelas 2 atau 3 dengan rata-rata 21.97 dan 19.74 serta 16.12 dan 12.66 berurut masing-masing perempuan dan laki-laki. Perbandingan survival rate antara kelas 1, 2, dan 3 sangat signifikan dengan angka 62.96%, 47.28%, dan 24.24%. Ini adalah contoh dari uang lebih penting daripada yang lain.